doc/man/man3/crypto_sign_init_first_pass_custom_hash.3monocypher

```diff
@@ -0,0 +1,234 @@
+.Fa init ,
+.Fa update
+and
+.Fa final
+members.
+.It Fa init
+Function that initialises the hash context an outer signing context.
+.It Fa update
+Function that updates the hash context of an outer signing context.
+It must be able to handle message sizes of at least 32 bytes.
+.It Fa final
+Function that finalises the hash context of an outer signing context.
+This function should wipe the hash context with
+.Xr crypto_wipe 3monocypher
+if it contains pointers to objects outside the outer signing context.
+Monocypher takes care of wiping the outer signing context.
+.It Fa ctx_size
+The size of the outer signing context as determined by
+.Fn sizeof .
+.El
+.Pp
+The functions indicated in the
+.Vt crypto_sign_vtable
+must be thread-safe if any of Monocypher's signing functions are
+accessed from multiple threads.
+.Pp
+After calling
+.Fn crypto_sign_init_first_pass_custom_hash
+or
+.Fn crypto_check_init_custom_hash ,
+the
+.Xr crypto_sign_update 3monocypher ,
+.Xr crypto_sign_final 3monocypher ,
+.Xr crypto_sign_init_second_pass 3monocypher ,
+.Xr crypto_check_update 3monocypher ,
+and
+.Xr crypto_check_final 3monocypher
+functions can be used as usual.
+They will call into the hash vtable as required.
+The same security considerations and semantics apply.
+.Sh RETURN VALUES
+These functions return nothing.
+.Sh EXAMPLES
+Defining and using a custom implementation of SHA-512 and crudely
+checking its results against
+.Xr crypto_ed25519_sign 3monocypher :
+.Bd -literal -offset indent
+struct outer_ctx {
+    crypto_sign_ctx_abstract sctx;
+    SHA2_CTX hash_ctx;
+};
+
+static void
+my_hash(uint8_t hash[64], const uint8_t *msg, size_t len)
+{
+    SHA2_CTX ctx;
+    SHA512Init(&ctx);
+    SHA512Update(&ctx, msg, len);
+    SHA512Final(hash, &ctx);
+}
+
+static void
+my_init(void *ctx)
+{
+    struct outer_ctx *octx = ctx;
+    SHA512Init(&octx->hash_ctx);
+}
+
+static void
+my_update(void *ctx, const uint8_t *msg, size_t len)
+{
+    struct outer_ctx *octx = ctx;
+    SHA512Update(&octx->hash_ctx, msg, len);
+}
+
+static void
+my_final(void *ctx, uint8_t *hash)
+{
+    struct outer_ctx *octx = ctx;
+    SHA512Final(hash, &octx->hash_ctx);
+}
+
+static const crypto_sign_vtable my_vtable = {
+    my_hash,
+    my_init,
+    my_update,
+    my_final,
+    sizeof(struct outer_ctx)
+};
+
+int
+main(void)
+{
+    uint8_t theirs[64], mine[64];
+    uint8_t sk[32];
+    const uint8_t msg[] = {
+        0x00, 0x01, 0x02, 0x04
+    };
+
+    arc4random_buf(sk, sizeof(sk));
+    crypto_ed25519_sign(theirs, sk, NULL, msg, sizeof(msg));
+
+    struct outer_ctx ctx;
+    crypto_sign_init_first_pass_custom_hash((void*)&ctx,
+                                            sk, NULL, &my_vtable);
+    crypto_sign_update(          (void*)&ctx, msg, sizeof(msg));
+    crypto_sign_init_second_pass((void*)&ctx);
+    crypto_sign_update(          (void*)&ctx, msg, sizeof(msg));
+    crypto_sign_final(           (void*)&ctx, mine);
+
```

Comment:
You may want to wipe the secret key, unless you think that'd clutter the example (the context itself is wiped, so no need to do that explicitly).